src/libstore/build/substitution-goal.cc

```diff
@@ -102,29 +94,42 @@ void PathSubstitutionGoal::tryNext()
         if (sub->storeDir == worker.store.storeDir)
             assert(subPath == storePath);
     } else if (sub->storeDir != worker.store.storeDir) {
-        tryNext();
-        return;
+        co_return tryNext();
     }
 
+    // Horrible, horrible code.
+    // Needed because we can't `co_*` inside a catch-clause.
+    // `std::variant` would be cleaner perhaps.
```

Comment:
I image you can soon convert this stuff into a simpler for loop like you did with the other substitution goal?